File: 0202.HappyNumber.py
class Solution(object):
    def isHappy(self, n):
        """
        :type n: int
        :rtype: bool
        """
        if n <= 0:
            return False

        num = n
        res = []
        while True:
            add = 0
            while num > 0:
                add += (num%10) ** 2
                num = num / 10
            if add == 1:
                return True
            else:
                for i in range(len(res)):
                    if add == res[i]:
                        return False
            res.append(add)
            num = add
            print res

test = Solution()
print test.isHappy(3)
'''
for i in range(15):
    print i, test.isHappy(i)
'''

File: 0202.HappyNumber.go
func isHappy(n int) bool {
    Map := make(map[int]bool)
    for n != 1 {
        if Map[n] {
            return false
        }
        Map[n] = true
        n = square(n)
    }
    return true
}

func square(n int) int {
    sq := 0
    for n > 0 {
        d := n%10
        sq += d*d
        n /= 10
    }
    return sq
}